### ISSUE TITLE
FontAwesome nicht in /assets

### ISSUE THREAD

**Issue by bizon1** [netzmacht/contao-font-awesome#22]
**Using config.component-dir has been deprecated. Please use extra.contao-component-dir instead.**
Wenn man dieser Meldung beim composer update folgt und die composer.json entsprechend ändert, dann landet die Extension netzmacht/contao-font-awesome bei der Installation in Contao-4.5.21 und Contao-4.6-dev nicht mehr im Ordner /assets/font-awesome/ sondern eine Ebene höher in /components/font-awesome/.
Dadurch funktionieren die Fonts nicht mehr.

Wenn man o.g. Meldung ignoriert und die composer.json also nicht geändert, funktioniert alles.

**Comment by dmolineus** [comment]
Der Fehler liegt hier eher bei https://github.com/RobLoach/component-installer

Da das Package nicht mehr weiterentwickelt wird, habe ich keine Ahnung wie man dies fixen kann ohne einen BC Break herbeizuführen. Die vorgeschlagenen Alternativen [Asset Packagist](https://asset-packagist.org/) or [Component Installers Extender](https://github.com/oomphinc/composer-installers-extender) halte ich nicht für optimal für die Erweiterung

**Comment by bizon1** [comment]
Danke für die schnelle Antwort.

Ich habe es vorerst so behoben, indem ich einen Symlink auf /components/font-awesome in /assets gelegt habe.

**Comment by Stefko** [comment]
> Ich habe es vorerst so behoben, indem ich einen Symlink auf /components/font-awesome in /assets gelegt habe.

Habe das selbe Problem, könntest Du bitte kurz erklären, wie Du das gemacht hast? Danke :-)

**Comment by fritzmg** [comment]
> Der Fehler liegt hier eher bei https://github.com/RobLoach/component-installer

Eigentlich nicht, das ist schon korrekt so. Wenn kein `component-dir` in der root composer.json angegeben ist, werden die components by default in den `components` Ordner gegeben.

Die contao-font-awesome Extension müsste daher prüfen, welcher dieser Ordner in Verwendung ist.

@Stefko: du musst einfach die `component-dir` konfigurieren:
```
    "config": {
        "component-dir": "assets"
    },
    "extra": {
        "contao-component-dir": "assets"
    },
```

**Comment by dmolineus** [comment]
> Eigentlich nicht, das ist schon korrekt so. 

Stimmt, ich habe da zwei Sachen durcheinander gebracht. Bisher nutzten Contao und der Component-Installer beide config.component-dir. Contao nun nicht mehr. Die Deprecated-Warning muss also ignoriert werden.

@fritzmg Problem an einen nicht gesetzten component-dir ist aber, dass die Stylesheets dann nicht geladen würden, oder übersehe ich da gerade etwas?

**Comment by fritzmg** [comment]
Hm ja stimmt, das `component/` Verzeichnis würde ja dann nicht symlinked werden. Könntest du aber selbst einbauen.

**Comment by fritzmg** [comment]
Vielleicht ist es aber ohnehin besser, du bundlest die assets in deinem package mit (evt. automatisiert).

**Comment by dmolineus** [comment]
> Vielleicht ist es aber ohnehin besser, du bundlest die assets in deinem package mit (evt. automatisiert).

Davon möchte ich eigentlich wegkommen. Mit den derzeitigen Stand, dass ich für jedes Release die Icon-Informationen aktualisieren muss, ist es egal. Ziel ist es aber, dass ich diese automatisiert auslesen kann (https://github.com/FortAwesome/Font-Awesome/issues/5867).

**Comment by fritzmg** [comment]
Naja, das Problem ist, dass du so oder so ein shim repository über composer braucht - zB eben components/font-awesome. Ich hätte auch gern eine bessere Lösung für solche assets, aber der aktuelle Stand der Dinge mit solchen assets über composer ist eher meh.

**Comment by dmolineus** [comment]
Ja, die Abhängigkeit besteht. Ich würde aber gern dem Nutzer die Entscheidung überlassen, welche Version er verwendet. Und selbst nicht immer hinterher sein zu müssen das Package zu aktualisieren.

@fritzmg Hast du zufällig Erfahrung gesammelt mit https://asset-packagist.org/?

**Comment by fritzmg** [comment]
> @fritzmg Hast du zufällig Erfahrung gesammelt mit https://asset-packagist.org/?

Noch nicht - aber solche Lösungen habe ich bisher nicht für Contao Extensions beachtet, wenn sie Änderungen in der root composer.json benötigen (wie hier zB das zusätzliche repository).

Hast du schon mal https://github.com/oomphinc/composer-installers-extender probiert? Das sollte ja ohne Änderung der root composer.json funktionieren, wenn ich mich nicht täusche.

**Comment by dmolineus** [comment]
Noch nicht, aber es benötigt auch eine Konfiguration unter extra in der Root-Konfiguration, siehe https://github.com/oomphinc/composer-installers-extender#how-to-use Und man bliebe auf den Shim-Repositories. Im Prinzip ist es ja nichts anderes als ein Ersatz zu den robloach/components-installer. Dieser benötigt halt auch den Eintrag in der composer.json - den Contao damals glücklicherweise übernommen hatte.

**Comment by fritzmg** [comment]
> Noch nicht, aber es benötigt auch eine Konfiguration unter extra in der Root-Konfiguration, siehe https://github.com/oomphinc/composer-installers-extender#how-to-use Und man bliebe auf den Shim-Repositories.

Hm, sicher? "extra" Einträge könnten - je nach dem, auch in den packages funktionieren.

**Comment by dmolineus** [comment]
> Hm, sicher? "extra" Einträge könnten - je nach dem, auch in den packages funktionieren.

Ja, hatte den Code gecheckt, weil ich die Readme nicht eindeutig fand.

**Comment by dive13** [comment]
Wie wäre jetzt die konkrete Lösung für das Problem? Welche Schritte müssen gemacht werden, damit man in einer Contao 4.6 wieder auf die icons zugreifen kann? In meiner composer.json steht bereits:
`    "extra": {
        "contao-component-dir": "assets"
    },`

**Comment by dmolineus** [comment]
> Wie wäre jetzt die konkrete Lösung für das Problem? Welche Schritte müssen gemacht werden, damit man in einer Contao 4.6 wieder auf die icons zugreifen kann? In meiner composer.json steht bereits:
> `"extra": { "contao-component-dir": "assets" },`

Du benötigst weiterhin den Eintrag unter `config`. Die deprecated-Warning von Contao innerhalb von Composer kannst du ignorieren.

```json
    "config": {
        "component-dir": "assets"
    }
```

**Comment by dive13** [comment]
… verstehe, danke.

**Comment by designpilot** [comment]
@dmolineus 
Wenn ich das so in der **composer.json** drin habe funktioniert das aber auch nicht

```
     "config": {
         "component-dir": "assets"
     },
    "extra": {
        "contao-component-dir": "assets"
    },
```

weil **a)** das hier fehlt **/assets/font-awesome/css/** und **b)** ich das mit dem Contao Manager unter Systemwartung alles ausgeführt habe und auch die Pakete aktualisiert habe, es hat alles nichts geholfen, auch nicht das ich Font Awesome deinstalliert habe und nochmals neu installiert hab. Dabei habe ich auch nach der Deinstallation von Font Awesome die oben stehenden Zeilen in der composer.json stehen lassen und nochmal probiert - ohne Erfolg - Der Fehler wird immer angezeigt und das schlägt mit einem Eintrag in der **/var/log/** zu buche.

**Comment by Aybee** [comment]
Ich wollte das mal pushen, damit das nach der Installation wieder funktioniert.

**Comment by kroerig** [comment]
kommt hier eigentlich noch eine Version, die wieder out-of-the-box für jedermann nutzbar ist?

**Comment by dmolineus** [comment]
> kommt hier eigentlich noch eine Version, die wieder out-of-the-box für jedermann nutzbar ist?

Ist derzeit nicht geplant. Die Lösungen müssen derzeit in der root composer.json vorgenommen werden. Die Erweiterung unterstützt ja aktuell nur Font Awesome 4, in eigenen Projekten bin ich seit einiger Zeit auf Font Awesome 5 umgestiegen. Der Bedarf an dieser Erweiterung meinerseits ist nicht mehr vorhanden. 

Selbstverständlich können entsprechende Features/Wünsche auch beauftragt werden.
